File: deno/lib/validators/email.ts
import {
  ServiceException,
} from "/lib/exceptions/service.exception.ts";

import {
  ns,
} from "/src/base/i18n/i18n.ts";

const REG = /^[a-zA-Z0-9_\-\.]+@[a-zA-Z0-9_\-\.]+$/gm;

export async function email(
  value?: string | null,
  label?: string,
) {
  if (value == null) {
    return;
  }
  if (REG.test(value)) {
    return;
  }
  label = label || "";
  const err_msg = await ns(
    "邮件格式不正确",
  );
  throw new ServiceException(`${ label } ${ err_msg }`, "validate_email");
}

File: rust/src/common/validators/email.rs
use anyhow::Result;
use fast_chemail::is_valid_email;

use crate::common::context::Ctx;
use crate::common::context::SrvErr;
use crate::src::base::i18n::i18n_dao::ns;

#[allow(dead_code)]
pub async fn email(
  ctx: &mut Ctx,
  value: Option<String>,
  label: &str,
) -> Result<()> {
  
  if value.is_none() {
    return Ok(());
  }
  let value = value.unwrap();
  
  if is_valid_email(&value) {
    return Ok(());
  }
  
  let msg = ns(
    ctx,
    "邮件格式不正确".to_owned(),
    None,
  ).await?;
  
  let mut err_msg = String::new();
  err_msg.push_str(label);
  err_msg.push(' ');
  err_msg.push_str(&msg);
  let err_msg = err_msg;
  
  Err(SrvErr::msg(err_msg).into())
}
